code/modules/flufftext/Hallucination.dm

```diff
@@ -346,82 +397,122 @@ GLOBAL_LIST_INIT(hallucination_list, list(
 		if("harmbaton") //zap n slap
+			iterations_left = rand(5, 12)
 			target.playsound_local(source, 'sound/weapons/egloves.ogg', 40, 1)
 			target.playsound_local(source, get_sfx("bodyfall"), 25, 1)
-			sleep(20)
-			for(var/i in 1 to rand(5, 12))
-				target.playsound_local(source, "swing_hit", 50, 1)
-				sleep(rand(CLICK_CD_MELEE, CLICK_CD_MELEE + 4))
+			next_action = 2 SECONDS
 		if("bomb") // Tick Tock
-			for(var/i in 1 to rand(3, 11))
+			iterations_left = rand(3, 11)
+
+	if (process)
+		START_PROCESSING(SSfastprocess, src)
+	else
+		qdel(src)
+
+/datum/hallucination/battle/process(delta_time)
+	next_action -= (delta_time * 10)
+
+	if (next_action <= 0)
+		switch (battle_type)
+			if ("disabler", "laser", "gun")
+				var/fire_sound
+				var/hit_person_sound
+				var/hit_wall_sound
+				var/number_of_hits
+				var/chance_to_fall
+
+				switch (battle_type)
+					if ("disabler")
+						fire_sound = 'sound/weapons/taser2.ogg'
+						hit_person_sound = 'sound/weapons/tap.ogg'
+						hit_wall_sound = 'sound/weapons/effects/searwall.ogg'
+						number_of_hits = 3
+						chance_to_fall = 70
+					if ("laser")
+						fire_sound = 'sound/weapons/laser.ogg'
+						hit_person_sound = 'sound/weapons/sear.ogg'
+						hit_wall_sound = 'sound/weapons/effects/searwall.ogg'
+						number_of_hits = 4
+						chance_to_fall = 70
+					if ("gun")
+						fire_sound = 'sound/weapons/gun/shotgun/shot.ogg'
+						hit_person_sound = 'sound/weapons/pierce.ogg'
+						hit_wall_sound = "ricochet"
+						number_of_hits = 2
+						chance_to_fall = 80
+
+				target.playsound_local(source, fire_sound, 25, 1)
+
+				if(prob(50))
+					addtimer(CALLBACK(target, /mob/.proc/playsound_local, source, hit_person_sound, 25, 1), rand(5,10))
+					hits += 1
+				else
+					addtimer(CALLBACK(target, /mob/.proc/playsound_local, source, hit_wall_sound, 25, 1), rand(5,10))
+
+				next_action = rand(CLICK_CD_RANGE, CLICK_CD_RANGE + 6)
+
+				if(hits >= number_of_hits && prob(chance_to_fall))
+					addtimer(CALLBACK(target, /mob/.proc/playsound_local, source, get_sfx("bodyfall"), 25, 1), next_action)
+					qdel(src)
+					return
+			if ("esword")
+				target.playsound_local(source, 'sound/weapons/blade1.ogg', 50, 1)
+
+				if (hits == 4)
+					target.playsound_local(source, get_sfx("bodyfall"), 25, 1)
+
+				next_action = rand(CLICK_CD_MELEE, CLICK_CD_MELEE + 6)
+				hits += 1
+
+				if (iterations_left == 0)
+					target.playsound_local(source, 'sound/weapons/saberoff.ogg', 15, 1)
+			if ("harmbaton")
+				target.playsound_local(source, "swing_hit", 50, 1)
+				next_action = rand(CLICK_CD_MELEE, CLICK_CD_MELEE + 4)
+			if ("bomb")
 				target.playsound_local(source, 'sound/items/timer.ogg', 25, 0)
-				sleep(15)
-	qdel(src)
+				next_action = 15
+
+		iterations_left -= 1
+		if (iterations_left == 0)
+			qdel(src)
+
+/datum/hallucination/battle/Destroy()
+	. = ..()
+	STOP_PROCESSING(SSfastprocess, src)
```

Comment:
You don't actually need to clean up turf refs, but since you do so in a previous destroy, I assume you forgot about it